plugins/nf-co2footprint/src/main/nextflow/co2footprint/Outfiles/CO2FootprintReport.groovy

```diff
@@ -0,0 +1,224 @@
+package nextflow.co2footprint.Outfiles
+
+import nextflow.co2footprint.CO2EquivalencesRecord
+import nextflow.co2footprint.CO2FootprintConfig
+import nextflow.co2footprint.CO2FootprintResourcesAggregator
+import nextflow.co2footprint.CO2Record
+import nextflow.co2footprint.utils.Converter
+
+import groovy.text.GStringTemplateEngine
+import groovy.text.Template
+import groovy.util.logging.Slf4j
+
+import nextflow.Session
+import nextflow.processor.TaskId
+import nextflow.trace.TraceHelper
+import nextflow.trace.TraceRecord
+
+import java.nio.file.Path
+
+
+@Slf4j
+/**
+ * Class to generate the Text file
+ */
+class CO2FootprintReport extends CO2FootprintFile{
+
+    /**
+     * Maximum number of tasks until table is dropped
+     */
+    private int maxTasks
+
+    // Information for final report
+    private Double total_energy
+    private Double total_co2
+    private CO2EquivalencesRecord equivalences
+    private CO2FootprintResourcesAggregator aggregator
+    private CO2FootprintConfig config
+    private String version
+    private Session session
+    private Map<TaskId, TraceRecord> traceRecords
+    private Map<TaskId, CO2Record> co2eRecords
+
+    CO2FootprintReport(Path path, boolean overwrite, int maxTasks) {
+        super(path, overwrite)
+        this.maxTasks = maxTasks
+    }
+
+    /**
+     * Write the final report and close the file
+     *
+     * @param total_energy
+     * @param total_co2
+     * @param equivalences
+     * @param aggregator
+     * @param config
+     * @param version
+     * @param session
+     * @param traceRecords
+     * @param co2eRecords
+     */
+    void close(
```

Comment:
Why did you move the actual output generation to `close()`? Not sure, I would expect this to be done explicitly before. At least that would make the functionalities more clear.

Reply:
Option 1 sounds good to me too!